pkg/bucket/replication/filter.go

```diff
@@ -0,0 +1,124 @@
+/*
+ * MinIO Cloud Storage, (C) 2020 MinIO, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package replication
+
+import (
+	"encoding/xml"
+)
+
+var (
+	errInvalidFilter        = Errorf("Filter must have exactly one of Prefix, Tag, or And specified")
+	errMissingOrEmptyFilter = Errorf("Filter missing or empty")
+)
+
+// Filter - a filter for a replication configuration Rule.
+type Filter struct {
+	XMLName xml.Name `xml:"Filter"`
+	Prefix  string
+	And     And
+	Tag     Tag
+	// Caching tags, only once
+	cachedTags []string
```

Comment:
```suggestion
	cachedTags map[string]struct{}
```

This will allow to do O(1) lookups of tags.

Reply:
good idea, changed this.